src/main/java/com/amazon/ion/impl/bin/IonRawBinaryWriter.java

```diff
@@ -251,14 +248,50 @@ public ContainerInfo()
             patchIndex = -1;
         }
 
+        /**
+         * Set the data in the patch point assigned to this container.
+         * 
+         * PatchPoint instances in {@link #patchPoints} are reused across the lifecycle of the writer, so
+         * there is a chance that a patch point already exists at {@link #patchIndex} with stale data
+         * that needs to be replaced. However, PatchPoints are not constructed until their first use,
+         * meaning the slot pointed to by {@link #patchIndex} may be null, in which case we have to
+         * create our own. This can occur if a child container ends up needing a patch point and some
+         * of its ancestors are missing patch points; the parents are assigned indices into {@link #patchPoints}
+         * but the patch point object itself was not constructed until now.
+         */
+        private void setPatchPointData(final long oldPosition, final int oldLength, final long length)
+        {
+            final PatchPoint existingPatchPoint = patchPoints.get(patchIndex);
+            if (existingPatchPoint != null) {
+                existingPatchPoint.initialize(oldPosition, oldLength, length);
+            } else {
+                patchPoints.set(patchIndex, new PatchPoint().initialize(oldPosition, oldLength, length));
+            }
+        }
+
         public void appendPatch(final long oldPosition, final int oldLength, final long length)
         {
             if (patchIndex == -1) {
                 // We have no assigned patch point, we need to make our own
-                patchIndex = patchPoints.push(p -> p.initialize(oldPosition, oldLength, length));
+                patchPointsLength++;
+                patchIndex = patchPointsLength - 1;
+                if (patchIndex < patchPoints.size()) {
+                    // The patch point ArrayList is large enough, the index of this patch point is in bounds.
+                    // A stale reusable patch point instance may be waiting for us here, or this may be a null slot.
+                    setPatchPointData(oldPosition, oldLength, length);
+                } else {
+                    // The patch point ArrayList is not large enough. It needs to grow to accomodate this
+                    // patch point. No need to call setPatchPointData since we know the PatchPoint instance
+                    // is missing.
+                    patchPoints.ensureCapacity(patchPointsLength);
+                    for (int i = patchPoints.size(); i < patchPointsLength - 1; i++) {
+                        patchPoints.add(null);
+                    }
+                    patchPoints.add(new PatchPoint().initialize(oldPosition, oldLength, length));
+                }                
```

Comment:
This costs one extra `add(null)` and `get(index)` in the case that we're growing the array, but is otherwise more straightforward to reason about- we always use `setPathPointData`. Your mileage may vary, you don't have to take this one. 

```suggestion
                if (patchIndex >= patchPoints.size()) {
                    // The patch point ArrayList is not large enough. It needs to grow to accomodate this
                    // patch point. No need to call setPatchPointData since we know the PatchPoint instance
                    // is missing.
                    patchPoints.ensureCapacity(patchPointsLength);
                    for (int i = patchPoints.size(); i < patchPointsLength ; i++) {
                        patchPoints.add(null);
                    }
                } 

                // A stale reusable patch point instance may be waiting for us here, or this may be a null slot.
                setPatchPointData(oldPosition, oldLength, length);            
```